### ISSUE TITLE
Roslyn doesn't warn about incorrect `[DoesNotReturn]` on async methods

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#43251]
**Version Used**: 

```
commit ff3d2a75dcd9ad286d4bad47c993454f3d1dd1d7
Merge: 7be257abe9 6fd1f755aa
Author: msftbot[bot] <48340428+msftbot[bot]@users.noreply.github.com>
Date:   Thu Apr 2 15:49:03 2020 +0000

    Merge pull request #43000 from mavasani/PortConvertAnonymousTypeToTuple

    Port ConvertAnonymousTypeToTuple to shared layer
```

**Steps to Reproduce**:

Compile and run the following code
```cs
#nullable enable
using System.Threading.Tasks;
using System.Diagnostics.CodeAnalysis;
class C
{
  [DoesNotReturn] public static async Task M() { await Task.Delay(1000); while (true) { }; }
    
  public static void Main()
  {
    string? s = null;
    if (s == null)
    {
      var task = M();
    }
        
    s.ToString();
  }
}
```

**Expected Behavior**:
`Warning [CS8763]: A method marked [DoesNotReturn] should not return.`

**Actual Behavior**:
No warnings. The program crashes at runtime with a `NullReferenceException`

**Comment by RikkiGibson** [comment]
So we are expected to warn on the 'await'?

**Comment by TessenR** [comment]
That would be correct. I personally don't think it's worth tracking awaits and think warning for any async method with `[DoesNotReturn]` would suffice but you might want to be more rigorous about it
